src/storage/sftp/storage.c

```diff
@@ -0,0 +1,931 @@
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(file != NULL);
+
+    // Attempt to unlink the file
+    int rc;
+    Wait *const wait = waitNew(this->timeoutConnect);
+
+    do
+    {
+        rc = libssh2_sftp_unlink_ex(this->sftpSession, strZ(file), (unsigned int)strSize(file));
+    }
+    while (rc == LIBSSH2_ERROR_EAGAIN && waitMore(wait));
+
+    if (rc != 0)
+    {
+        if (rc == LIBSSH2_ERROR_SFTP_PROTOCOL)
+        {
+            if (param.errorOnMissing || !storageSftpLibSsh2FxNoSuchFile(this, rc))
+                storageSftpEvalLibSsh2Error(
+                    rc, libssh2_sftp_last_error(this->sftpSession), &FileRemoveError,
+                    strNewFmt("unable to remove '%s'", strZ(file)),
+                    NULL);
+        }
+        else
+        {
+            if (param.errorOnMissing)
+                THROW_FMT(FileRemoveError, "unable to remove '%s'", strZ(file));
+        }
+    }
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/**********************************************************************************************************************************/
+static StorageRead *
+storageSftpNewRead(THIS_VOID, const String *const file, const bool ignoreMissing, const StorageInterfaceNewReadParam param)
+{
+    THIS(StorageSftp);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_SFTP, this);
+        FUNCTION_LOG_PARAM(STRING, file);
+        FUNCTION_LOG_PARAM(BOOL, ignoreMissing);
+        FUNCTION_LOG_PARAM(UINT64, param.offset);
+        FUNCTION_LOG_PARAM(VARIANT, param.limit);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(file != NULL);
+
+    FUNCTION_LOG_RETURN(
+        STORAGE_READ,
+        storageReadSftpNew(
+            this, file, ignoreMissing, this->ioSession, this->session, this->sftpSession, this->sftpHandle, this->timeoutSession,
+            this->timeoutConnect, param.offset, param.limit));
+}
+
+/**********************************************************************************************************************************/
+static StorageWrite *
+storageSftpNewWrite(THIS_VOID, const String *const file, const StorageInterfaceNewWriteParam param)
+{
+    THIS(StorageSftp);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_SFTP, this);
+        FUNCTION_LOG_PARAM(STRING, file);
+        FUNCTION_LOG_PARAM(MODE, param.modeFile);
+        FUNCTION_LOG_PARAM(MODE, param.modePath);
+        FUNCTION_LOG_PARAM(STRING, param.user);
+        FUNCTION_LOG_PARAM(STRING, param.group);
+        FUNCTION_LOG_PARAM(TIME, param.timeModified);
+        FUNCTION_LOG_PARAM(BOOL, param.createPath);
+        FUNCTION_LOG_PARAM(BOOL, param.syncFile);
+        FUNCTION_LOG_PARAM(BOOL, param.syncPath);
+        FUNCTION_LOG_PARAM(BOOL, param.atomic);
+        FUNCTION_LOG_PARAM(BOOL, param.truncate);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(file != NULL);
```

Comment:
Because this driver will only be used for repos, it should have the following asserts (like S3, etc):
```
    ASSERT(param.createPath);
    ASSERT(param.truncate);
    ASSERT(param.user == NULL);
    ASSERT(param.group == NULL);
    ASSERT(param.timeModified == 0);
```
This is why the user/group/timestamp code is not reachable.

Reply:
added missing asserts,  updated tests accordingly commit f25431f5b67377280f90996e8aead476d88c2b6a